x11-utils.c

```diff
@@ -0,0 +1,227 @@
+#include <stdint.h>
+#include <unistd.h>
+#include <string.h>
+
+#include "x11-utils.h"
+#include "mce-log.h"
+
+static Atom x11_atom_touchscreen = None;
+static Atom x11_atom_device_enabled = None;
+static Atom x11_atom_device_enabled_type = None;
+static int x11_atom_device_enabled_format = 0;
+
+static Display* x11_get_display(void)
+{
+		Display* dpy = XOpenDisplay(NULL);
+		if (dpy == NULL) {
+			dpy = XOpenDisplay(":0.0");
+		}
+
+		if (dpy == NULL) {
+			mce_log(LL_INFO, "%s: unable to open display", __func__);
+		}
+		return dpy;
+}
+
+bool x11_set_input_device_enabled(Display* dpy, const XIDeviceInfo devinfo, const bool enable)
+{
+	if (dpy == NULL) {
+		mce_log(LL_WARN, "%s: XIDeviceInfo needs to be from same Display", __func__);
+		return false;
+	}
+	if (x11_atom_device_enabled == None)
+		x11_atom_device_enabled = XInternAtom(dpy, "Device Enabled", False);
+	if (x11_atom_device_enabled == None) {
+		mce_log(LL_WARN, "%s: unable to obtain X11 Atoms", __func__);
+		return false;
+	}
+	
+	if ((x11_atom_device_enabled_type == None) || (x11_atom_device_enabled_format == 0)) {
+		unsigned long ignore_bytes_after, ignore_nitems;
+		unsigned char *ignore_data = NULL;
+
+		if (XIGetProperty(dpy, devinfo.deviceid, x11_atom_device_enabled, 0, 0, False, AnyPropertyType, &x11_atom_device_enabled_type, &x11_atom_device_enabled_format, &ignore_nitems, &ignore_bytes_after, &ignore_data)) {
+			mce_log(LL_WARN, "%s: unable to obtain X11 Device Enabled property atom type", __func__);
+			return false;
+		} else {
+			XFree(ignore_data);
+		}
+	}
+	
+	XIChangeProperty(dpy, devinfo.deviceid, x11_atom_device_enabled,
+			x11_atom_device_enabled_type,
+			x11_atom_device_enabled_format, PropModeReplace, (unsigned char*)&enable, 1);
+	return true;
+}
+
+bool x11_set_all_input_devices_enabled(Display* dpy, const bool enable)
+{
+	bool ret = false;
+	XIDeviceInfo *devinfo = NULL;
+	int ndev = 0;
+	
+	static XIDeviceInfo *disabledDevices = NULL;
+	static unsigned int disabledDevicesCount = 0;
+
+	bool ownsDisplay = false;
+	if (dpy == NULL) {
+		dpy = x11_get_display();
+		if (dpy == NULL)
+			return false;
+		ownsDisplay = true;
+	}
+
+	if (x11_atom_touchscreen == None)
+		x11_atom_touchscreen = XInternAtom(dpy, XI_TOUCHSCREEN, True);
+
+	if (x11_atom_touchscreen == None) {
+		mce_log(LL_WARN, "%s: unable to obtain X11 Atoms", __func__);
+		goto done;
+	}
+
+	devinfo = XIQueryDevice(dpy, XIAllDevices, &ndev);
+	if (devinfo == NULL)
+		goto done;
+	
+	if (enable && disabledDevices == NULL) {
+		mce_log(LL_WARN, "%s: this function only enables devices previously disabled by it", __func__);
+		goto doneXiFree;
+	} else if (!enable && disabledDevices != NULL) {
+		mce_log(LL_WARN, "%s: this function can only disable devices once before renableing them", __func__);
+		goto doneXiFree;
+	} else if (!enable && disabledDevices == NULL) {
+		disabledDevices = malloc(sizeof(*disabledDevices)*ndev);
+	}
+
+	if (!enable) for (int i = 0; i < ndev; ++i) {
```

Comment:
please, don;t do that, my eyes bleed :)

Reply:
i find this more readable than the alternative but ok its changed